wikipron/extract/jpn.py

```diff
@@ -0,0 +1,229 @@
+and across pages.
+
+This extraction function attempts to grab data from as many entries
+as possible, but may not always grab all potential data on a given entry page.
+To reduce the complexity of scraping the sometimes widely different Japanese
+entry pages, this extraction function attempts to target "pairs" of words
+and pronunciations. It thus only grabs a word if it can find a "local"
+pronunciation entry on the page, and only grabs a pronunciation if it can
+find a "local" word on the page.
+
+The reason for not indiscriminately scraping each entry page for words or
+pronunciations that share a class is because of pages like:
+https://en.wiktionary.org/wiki/%E8%84%9A#Japanese
+https://en.wiktionary.org/wiki/%E5%B9%B3%E5%9C%B0#Japanese
+Where not all "Etymologies" contain pronunciations and not all "Etymologies"
+(as in the first link) are listed as a missing or incomplete. Scraping
+indiscriminately may lead to matching words and pronunciations that are not
+meant to be together.
+"""
+
+import itertools
+import typing
+
+import requests
+
+from wikipron.extract.default import yield_pron, IPA_XPATH_SELECTOR
+
+if typing.TYPE_CHECKING:
+    from wikipron.config import Config
+    from wikipron.typing import Iterator, Pron, Word, WordPronPair
+
+
+_HIRAGANA_WORD_XPATH_SELECTOR = """
+//b[contains(@class, "Jpan form-of lang-ja kana")]
+"""
+
+# Will also catch some purely hiragana entries and
+# hiragana entries that do not use the above HTML despite
+# being in the same position as them on the page.
+_KATAKANA_WORD_XPATH_SELECTOR = """
+//strong[@class = "Jpan headword" and @lang = "ja"]
+"""
+
+# Works upward from the word to its pronunciation
+_PRON_XPATH_SELECTOR = """
+{word_to_work_from}/..
+        /preceding-sibling::*[1][self::{heading}]
+            /preceding-sibling::*[1][self::ul]
+                {second_ul}
+"""
+
+# Assumes never more than two <ul>'s containing IPA pronunciations.
+# There may be an intervening <table>, <div>, or <h4>
+# between the "Pronunciation" heading and the word.
+# The selector as written does not handle skipping those.
+_PAIR_CHECK_XPATH_SELECTOR = """
+{word_to_grab}[
+    ..
+        /preceding-sibling::*[1][self::{heading}]
+            /preceding-sibling::*[1][self::ul][
+                (. | preceding-sibling::*[1][self::ul])
+                /preceding-sibling::*[1][self::{heading}][
+                    span[contains(@id, "Pronunciation")]
+                ]
+            ]
+
+]
+"""
+
+# Check for more than one etymology for the same reason as in lat.py
+# If there is more than one etymology, the headings get shifted down a level.
+_TOC_ETYMOLOGY_XPATH_SELECTOR = """
+//a[@href = "#Japanese"]
+  /following-sibling::ul
+    /li
+      /a
+        [starts-with(@href, "#Etymology")]
+"""
+
+
+# Some pages may not have etymology section.
+def _check_etymologies(request: requests.Response):
+    count = 0
+    for a_element in request.html.xpath(_TOC_ETYMOLOGY_XPATH_SELECTOR):
+        count += 1
```

Comment:
Would `count = len(request.html.xpath(_TOC_ETYMOLOGY_XPATH_SELECTOR))` simply work? I can't remember if `request.html.xpath` already returns a materialized array of things that would allow `len`. If not, and if we're keeping this `for` loop, `a_element` could simply be `_` as it's not used.

Reply:
If it's not an actual `len`-able object the hack is:

    count = sum(1 for _ in request.html.xpath(...))